semantic/src/main/kotlin/tools/samt/semantic/SemanticModelReferenceResolver.kt

```diff
@@ -111,27 +111,53 @@ internal class SemanticModelReferenceResolver(
                                 typeNode = expression.base,
                                 fullNode = expression,
                             )
+                        } else {
+                            expression.reportError(controller) {
+                                message("List must have exactly one type argument")
+                                highlight("expected one type argument", expression.location)
+                            }
                         }
                     }
 
                     "Map" -> {
                         if (expression.arguments.size == 2) {
-                            return ResolvedTypeReference(
-                                type = MapType(
-                                    keyType = resolveAndLinkExpression(scope, expression.arguments[0]),
-                                    valueType = resolveAndLinkExpression(scope, expression.arguments[1]),
-                                    node = expression,
-                                ),
-                                typeNode = expression.base,
-                                fullNode = expression,
+                            val keyTypeNode = expression.arguments[0]
+                            val keyType = resolveAndLinkExpression(scope, keyTypeNode)
+                            val valueType = resolveAndLinkExpression(scope, expression.arguments[1])
+                            val mapType = MapType(
+                                keyType = keyType,
+                                valueType = valueType,
+                                node = expression,
                             )
+
+                            // error if keyType does not refer to string type
+                            if (keyType.type != StringType) {
+                                keyTypeNode.reportError(controller) {
+                                    message("Map key type must be String")
+                                    highlight("expected String", keyTypeNode.location)
+                                }
+                            } else {
+                                return ResolvedTypeReference(
+                                    type = mapType,
+                                    typeNode = expression.base,
+                                    fullNode = expression,
+                                )
+                            }
+                        } else {
+                            expression.reportError(controller) {
```

Comment:
Maybe overkill: Add a dedicated message if one generic is provided? (e.g. Use List<...> instead with a recommendation